### PR TITLE
Tic-Tac-Toe application

### LINKED ISSUE #109: TicTacToe smart contract
# Summary

The game is simple in its original state which consists of

- 3x3 grid
- 2 players
- Each player makes a move in a turn based fashion
- First player to reach 3 in a row wins otherwise it's a draw

Expansions to game logic (more players, larger boards, more / less markers required to win etc.) and contract data handling (multiple games at once, allowing contracts to interact with the game etc.) can be added later.

## Rough Contract Design

- The contract should support 2 players and they should not be contracts
    - The creator of the game can specify themeselves as both players
- After a player makes a move the next move is automatically set to be made by the next player unless the game has ended
- The game ends in any of the following states
    - Either player has reached 3 in a row
        - This can be in any direction (horizontal, vertical or diagonal)
    - The grid is full and no more moves can be made
- A player cannot place their marker in a position that is already filled
- A player cannot make a move outside of the designated 3x3 grid
- There can only be 1 game played at a time
    - This means that the game itself is stored in storage
    - No other storage is required
- Starting a new game overwrites any previous game even if that game has not ended
- Events should be emitted when
    - A new game is created
        - Who the players are
    - A move is made
        - Which player has made a move and at which position
    - The game ends (draw or if someone won then who won)

### PR DESCRIPTION
**Pull request by Rubyt0x**
## Type of change

- New Feature

## Changes

This game has 2 functions: `new_game`and `make_move`.

You can:
- Create a game
- Make moves under certain conditions

Added:
- Testing
- Optimized imports and directory setup/formatting

## Notes

- You can play only one game at a time.

## Related Issues

fixes #109

### CODE REVIEW

**Review by Braqzen — CHANGES_REQUESTED** [review]
1. There are some files in this project that should not be added to the main repository. Could you remove them and make sure they do not get added to the commits? There is a special file that git uses to ignore anything that you do not want added into a commit.
    - `.DS_STORE`
    - `out` directory and all of its content
2. Copy and pasting the notes in the issue does not constitute a non-technical specification. Could you rework it a little to align closer to the non-technical spec in the [CONTRIBUTING.md](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#non-technical-specification) file?
    - Some of the content in the issue can be duplicated however some is a little too technical e.g. what is kept in "storage", events etc.
    - To help think about the spec consider how you would explain the game to a small child; what the game is, the rules etc.
3. Could you replicate the README content to be similar to the other applications e.g. the headings and the logo functionality so that it automatically changes based on the light/dark theme the user has set (and it has the name of the game next to the logo)? You can see an example in the other apps e.g. [here](https://raw.githubusercontent.com/FuelLabs/sway-applications/master/fundraiser/README.md)
    - The badges (CI passing, forc version and discord link) are not needed either because they are at the top level of the repository
4. There are some files that do not have an empty line at the end and github does not like that very much. Could you check the files and make sure that each file has an empty line at the end?
5. The `contract` and `player_identity` should really be in one directory since the `player_identity` is not really a "library" in the conventional sense. Could you rework it to be similar to the pattern in [CONTRIBUTING.md](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#project-structure) Here is a [link](https://github.com/FuelLabs/sway-applications/tree/master/fundraiser/src) to an app which uses the pattern.
6. Could you update the PR to use a special keyword to automatically close the issue? You can find more information [here](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#creating-the-pull-request)
7. Since this is a WIP this PR should really be a "draft", could you change the state of the PR? You can find some info about that [here](https://github.blog/2019-02-14-introducing-draft-pull-requests/) however that does not show you how to convert a PR into a draft. You can do that by clicking "Convert to draft" right under the "Reviewers" section.

After these changes I'll make some comments on the code itself. One of the things that will need to be done at the end is to bump up the rust SDK version to 0.16 and to write a bunch of tests to prove that the code works. Oh, and add it to the CI so that it validates.

**Comment by eureka-cpu** [comment]
> 1. There are some files in this project that should not be added to the main repository. Could you remove them and make sure they do not get added to the commits? There is a special file that git uses to ignore anything that you do not want added into a commit.
>        
>        * `.DS_STORE`
>        * `out` directory and all of its content
> 
>     2. Copy and pasting the notes in the issue does not constitute a non-technical specification. Could you rework it a little to align closer to the non-technical spec in the [CONTRIBUTING.md](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#non-technical-specification) file?
>        
>        * Some of the content in the issue can be duplicated however some is a little too technical e.g. what is kept in "storage", events etc.
>        * To help think about the spec consider how you would explain the game to a small child; what the game is, the rules etc.
> 
>     3. Could you replicate the README content to be similar to the other applications e.g. the headings and the logo functionality so that it automatically changes based on the light/dark theme the user has set (and it has the name of the game next to the logo)? You can see an example in the other apps e.g. [here](https://raw.githubusercontent.com/FuelLabs/sway-applications/master/fundraiser/README.md)
>        
>        * The badges (CI passing, forc version and discord link) are not needed either because they are at the top level of the repository
> 
>     4. There are some files that do not have an empty line at the end and github does not like that very much. Could you check the files and make sure that each file has an empty line at the end?
> 
>     5. The `contract` and `player_identity` should really be in one directory since the `player_identity` is not really a "library" in the conventional sense. Could you rework it to be similar to the pattern in [CONTRIBUTING.md](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#project-structure) Here is a [link](https://github.com/FuelLabs/sway-applications/tree/master/fundraiser/src) to an app which uses the pattern.
> 
>     6. Could you update the issue to use a special keyword to automatically close the issue? You can find more information [here](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#creating-the-pull-request)
> 
>     7. Since this is a WIP this PR should really be a "draft", could you change the state of the PR? You can find some info about that [here](https://github.blog/2019-02-14-introducing-draft-pull-requests/) however that does not show you how to convert a PR into a draft. You can do that by clicking "Convert to draft" right under the "Reviewers" section.
> 
> 
> After these changes I'll make some comments on the code itself. One of the things that will need to be done at the end is to bump up the rust SDK version to 0.16 and to write a bunch of tests to prove that the code works. Oh, and add it to the CI so that it validates.

is `TicTacToe` not a sufficient keyword? I'm not sure I understand based on the `contributing` what you mean by adding a new keyword.

**Comment by simonr0204** [comment]
Some thoughts on how to handle state:

Players are really just addresses. I'm not sure if there's any need to abstract them behind a Player type. Also, I think you can combine the notions of Game state and winner into one. A simpler way to structure your state could be something like :

```
// A game is either in progress, drawn, or won. If it is won, the winner address is stored.
pub enum State {
    InProgress: (),
    Drawn: (),
    Won: Address,
}

// A game is defined by the Players and the State.
pub struct Game {
    player1: Address,
    player2: Address,
    state: State,
}
```

To me this is a lot easier to reason about, although maybe I'm missing something in terms of what other stuff you want to implement for your `Player` type later on.

Also, do you want the contract to be able to host more than one game at a time? If so, you'll want the `map` and `player_turn` to be part of your game state. If not - do you need the `Game` abstraction at all ?

**Review by Braqzen — CHANGES_REQUESTED** [review]
Here are my follow up comments from the last review and some new ones

1. The `.DS_Store` files have not been removed. There's multiple instances throughout this PR
2. Similarly, the `out` directory is here as well
3. The top-level README is being altered beyond the scope of the PR. Could you undo those changes so that only the TicTacToe line is being added?
    - Furthermore, could you list the game as a bullet point under "Games" rather than as a top-level addition?
4. Could you make sure that all files end in one new line and no more than that? Some have too many and some not enough.
5. Could you point at a local "TicTacToe" logo rather than the light / dark themes for the overall repo?
6. Could you double check any text to make sure that the players are being referred to in a neutral sense i.e. no his/her etc.?
7. [Here](https://github.com/FuelLabs/sway-applications/blob/master/escrow/SPECIFICATION.md) is another example of a specification.
8. `contract` is a reserved keyword so a project cannot be called that
    - I believe I have mentioned that this project should be added to the CI in the previous PR. The CI would have caught this and threw an error. Can you add this project to CI?
9. `data_structures` should not be its own project and instead it should be part of the `contract` project. Could you move that into `contract` and use the `dep` keyword instead of adding it as an external library in Forc.toml?
10. Could you also bump the testing suite to use the latest version and add tests when the contract is ready to be tested?
    - A project should not be marked as open for review when it does not have any tests. You can request a review from any number of reviewers when the PR is still a draft
    - Could you mark the PR as a draft and leave it as a draft until you think it is ready to potentially be merged into master?
11. `main.sw`
    - There is an import using the glob operator (the * in `core::*`). This is a bad practice because it's unclear what is being imported and from where. All imports should be explicit. If it's not being used then can you remove it?
    - There are 3 imports from storage when only the `StorageMap` is needed since it implements the `get` and `store` behind the scenes
    - I believe I have mentioned this before. Could you take a look at the [CONTRIBUTING.md](https://github.com/FuelLabs/sway-applications/blob/master/CONTRIBUTING.md#code-quality) section for code quality? There are a lot of comments that can be mentioned here which are addressed there.
    - @simonr0204 has just made some comments about the data types and I have previously made similar-ish comments so I won't comment on that for now
    - There are a couple places where the player turn is checked using 2 `if` statements. Could you change them to be `if-else` statements?
    - `break` is now a keyword that can be used inside loops so there's no need to use a variable to track that anymore. Could you update your code to use the `break` keyword? [Here](https://github.com/FuelLabs/sway-applications/blob/a962e60f1afd9fad7535a801934ecbe052db2e0b/escrow/src/main.sw#L130-L157) is an example.
    - Could you remove the utility functions that are outside of the contract? Those lines of code can be used directly inside the contract without issues

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/contract/src/main.sw:15]: Which of these functions need to be in the ABI? Some of them look like internal functions that should only be called by other functions of this contract.

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/contract/src/main.sw:40]: Should this function check that the correct player is making their move, or are you doing this check somewhere else ?

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/contract/src/main.sw:91]: I don't know if wrapping the Sway syntax for stroage set/get adds anything in terms of readability here.

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/contract/src/main.sw:46]: Don't forget to call `next_player` at the end of this function to flip the turn counter!

File: games/TicTacToe/contract/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/contract/src/main.sw:40]: Good catch, I'll make the change.

File: games/TicTacToe/contract/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/contract/src/main.sw:91]: How else would you store the moves?

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/contract/src/main.sw:11]: It's worth considering how you're going to write an algorithm to efficiently calculate if there is a win on the board, when deciding how to represent the map. Something organized 3x3 will allow you to reason about indices directly.

File: games/TicTacToe/contract/src/main.sw
Comment (simonr0204) [reply games/TicTacToe/contract/src/main.sw:91]: I just mean, you can do `storage.map.insert(key, value);` directly in the function that needs to update the map, rather than abstracting the `insert` operation behind another function.

**Review by matt-user — COMMENTED** [review]
I'm taking over the frontend work from Austin for this application.  Currently the next step for the frontend is to connect it to the smart contract in this repo.  Before I do that though I would like to see some tests written to make sure the smart contract is working.  Also a couple questions I have:

1.) How is a winner decided?

2.) Can multiple games be played at once?

3.) How will we keep track of what players are in a game(s)?  This is necessary because the frontend will need to display the correct game(s) depending on what wallet it is currently connected to.  Maybe to make things easier a player can only participate in one game at a time.

**Comment by eureka-cpu** [comment]
Thanks for stepping in @matt-user !

File: games/TicTacToe/contract/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/contract/src/main.sw:20]: Good idea, will implement it

Comment (Rubyt0x) [reply games/TicTacToe/contract/src/main.sw:20]: The problem with that is that you can play only one game at the time

File: games/TicTacToe/contract/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/contract/src/main.sw:97]: The purpose of this function is to call it when conditions are verified such as `map_is_full` or when winning conditions (under development) are met

Comment (Braqzen) [reply games/TicTacToe/contract/src/main.sw:97]: Tip: This should not be a function that has to be called otherwise other functions may break game logic. This should automatically happen after the last move is made. This means that checks must be performed before a move it made to first validate the move and then to decide if the game has finished. Alternatively there are no more moves and nothing happens.

File: games/TicTacToe/Forc.toml
Comment (simonr0204) [inline games/TicTacToe/Forc.toml:2]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

File: games/TicTacToe/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/src/main.sw:38]: Is there anything to stop one of the players (or even a 3rd party!) ruining the game by calling `new_game` half way through and resetting everything ?

File: games/TicTacToe/src/main.sw
Comment (simonr0204) [inline games/TicTacToe/src/main.sw:32]: Should `player_turn` be an option? Is there a valid game state where the turn is undefined?

File: games/TicTacToe/src/main.sw
Comment (matt-user) [inline games/TicTacToe/src/main.sw:31]: this comment should be above the `move_counter` variable

File: games/TicTacToe/src/utils.sw
Comment (matt-user) [inline games/TicTacToe/src/utils.sw:16]: I would add a comment here explaining the purpose of this variable.  I understand it is all of the possible win conditions, but it is not immediately apparent.

File: games/TicTacToe/src/utils.sw
Comment (matt-user) [inline games/TicTacToe/src/utils.sw:44]: I'm wondering if it would be more efficient to put the repeated matches variable in storage.  Do you know @simonr0204 ?

**Review by matt-user — COMMENTED** [review]
It's coming together!  Although still needs tests

File: games/TicTacToe/src/utils.sw
Comment (simonr0204) [reply games/TicTacToe/src/utils.sw:44]: hmm so since this is a constant, it's probably better for it to be in the bytecode rather than in storage. Storage reads are $$$

File: games/TicTacToe/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/src/main.sw:38]: Under the `move` function, there is a condition that checks if the right player is playing, so that's reducing the risks of compromising a game, but yeah good call, I'll add a condition to make sure no external party can reset the game in the middle of it.

File: games/TicTacToe/src/main.sw
Comment (Rubyt0x) [reply games/TicTacToe/src/main.sw:32]: As of now, yes because we use the `Option::None` state to stop the `move` function if there is either a winner or a draw, but this will be fixed in the next updates.

Comment (Rubyt0x) [reply games/TicTacToe/src/main.sw:32]: As of now, yes because we use the `Option::None` state to stop the `move` function if there is either a winner or a draw, but this will be fixed in the next updates.

File: games/TicTacToe/src/main.sw
Comment (Braqzen) [reply games/TicTacToe/src/main.sw:38]: The assumption was that this will run locally for now and that no other party will reset it.

The problem with adding this check is that now you need to decide when a new game can be started.

- Is it fine to start a new game only once the previous game is over?
- Is it fine to start a new game once either of the players decides that it is over?

The first point is probably okay.

`new_game()` can be called either when it has some default pre-init value or another value which indicates that the game has ended. This way any address can call the function if-and-only-if (_iff_) there is no active game atm.

This leads to the next problem of general race conditions for who gets to play. First caller gets to decide the players and thus additional functionality now needs to be introduced to either juggle multiple games at once internally or we need access control per contract.

There is also a secondary problem of a player never making a move and thus the contract is locked since `new_game()` cannot be called because the game has not reached a conclusion. Now we need additional functionality through privileges, time spent on a game etc. before a new game can be started so that the contract is not locked.

File: games/TicTacToe/src/utils.sw
Comment (Braqzen) [reply games/TicTacToe/src/utils.sw:16]: We have considered factoring out the array into storage and using it as a `read-only` value; however, storage does not support regular arrays.
Alternatively, we've considered moving it into a function and using that as a getting in the utility functions.

**Comment by Braqzen** [comment]
> It's coming together! Although still needs tests

Yup! The idea was to get some eyes on this and see what people think and if that's good then move ahead with a test suite.

File: games/TicTacToe/src/main.sw
Comment (simonr0204) [reply games/TicTacToe/src/main.sw:38]: I think this and my comment below could be solved by explicitly storing the game's state. Then it's easy to see if a game is complete or not. 

If the current design is tailored to a "friendly" game, then the risk of a player abandoning the game is lower than a malicious 3rd party resetting it. We can add a chess clock mechanism in a later PR.

Happy to defer resolution of these attack vectors to a later PR, but we should document them somewhere. (Either TODOs or preferably issues)

Comment (Braqzen) [reply games/TicTacToe/src/main.sw:38]: What is the purpose of another storage variable?

We could use another variable; but, we can also iterate over the positions and figure out if the game has ended. With the gas estimation provided by the SDK we can check which implementation is cheaper. There's also code quality to consider. It's significantly cleaner to access a variable than to have a method to iterate and figure out a state.

**Comment by nfurfaro** [comment]
You don't want to commit any `.DS_Store` files.

File: games/TicTacToe/project/tictactoe-contract/src/utils.sw
Comment (Braqzen) [inline games/TicTacToe/project/tictactoe-contract/src/utils.sw:3]:
```suggestion
```

File: games/TicTacToe/project/tictactoe-contract/src/interface.sw
Comment (nfurfaro) [inline games/TicTacToe/project/tictactoe-contract/src/interface.sw:13]: Can we be more specific here? i.e: does the function revert if there is any game currently playing, or only if there's already an active game between these same two players?
Also, consider this: given 2 players `A` and `B`, a game where `player_one = A` & `player_two = B` is fundamentally different than a game where `player_one = B` & `player_two = A`. Maybe it's better left to a future feature addition, but supporting both scenarios in parallel is a possibiity.

**Comment by Braqzen** [comment]
> You don't want to commit any `.DS_Store` files.

@Rubyt0x In the root of the repository can you run `ls -la` and check if the `.DS_Store`

> You don't want to commit any `.DS_Store` files.

@Rubyt0x In the repository root can you run `ls -la` to check if the `.DS_Store` file shows up? It does on my system. If it shows up then you can run `rm -f .DS_Store` to remove it. If there's an issue then lmk and I can remove it since I can see it on my system.

File: games/TicTacToe/project/tictactoe-contract/src/main.sw
Comment (nfurfaro) [inline games/TicTacToe/project/tictactoe-contract/src/main.sw:41]: another consideration: perhaps we want to support more than one game concurrently, in which case we could store a `Game` struct which contains stuff like the players, the status, etc...

**Comment by Rubyt0x** [comment]
> Just fixed it

**Review by Braqzen — DISMISSED** [review]

**Comment by Braqzen** [comment]
I recommend waiting for at least 1 more reviewer to approve before merging.

**Review by nfurfaro — COMMENTED** [review]
Looking great Ruben!
I left a couple comments; in general I would look at ways to reduce the number of storage reads & writes :)

File: games/TicTacToe/project/tictactoe-contract/src/main.sw
Comment (Braqzen) [reply games/TicTacToe/project/tictactoe-contract/src/main.sw:41]: Yeap; however, for a first addition a simple game is fine. We can make improvements with further PRs

File: games/TicTacToe/project/tictactoe-contract/src/interface.sw
Comment (Braqzen) [reply games/TicTacToe/project/tictactoe-contract/src/interface.sw:13]: There can only be 1 game played at a time. Once the game ends, via a draw or win, then the `new_game()` function can be called again.

File: games/TicTacToe/SPECIFICATION.md
Comment (supiket) [inline games/TicTacToe/SPECIFICATION.md:38]:
```suggestion
2. Checks if the move ends the game with either one player winning or a draw.
3. Reverts:
   1. When the game has already ended.
   2. If the incorrect player tries to make a move.
   3. If a player tries to make a move out of bounds.
   4. If a player tries to make a move to an occupied cell.
```
Another suggestion is to remove all the dots from the line endings for the items in the "Use Cases"

File: games/TicTacToe/project/tictactoe-contract/Cargo.toml
Comment (supiket) [inline games/TicTacToe/project/tictactoe-contract/Cargo.toml:9]: Though this is by no means a strict requirement, I would like to point out that all apps use `fuels 0.34.0`

File: games/TicTacToe/project/tictactoe-contract/Forc.toml
Comment (supiket) [inline games/TicTacToe/project/tictactoe-contract/Forc.toml:5]:
```suggestion
name = "tictactoe-contract"
```
Applying this suggestion would require changing the paths in `tests/utils/setup.rs`

**Review by supiket — CHANGES_REQUESTED** [review]
Looks very cool! I have minor suggestions that can be left for the next iterations, except the one regarding `.gitignore`

File: games/TicTacToe/project/tictactoe-contract/Cargo.toml
Comment (Braqzen) [reply games/TicTacToe/project/tictactoe-contract/Cargo.toml:9]: Oh right, I've mentioned that in our meet but I've forgotten about this

**Review by Braqzen — APPROVED** [review]

**Review by supiket — APPROVED** [review]

**Pull request merged**
